File: forge_engine/rl/actions.py
"""Action space definitions for RL environments.

Each ActionSpace defines:
  1. The gymnasium space (discrete or continuous)
  2. How to translate an RL action into engine trading commands
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple

import numpy as np
import gymnasium as gym


class ActionSpace(ABC):
    """Base class for action spaces."""

    @abstractmethod
    def gymnasium_space(self) -> gym.Space:
        """Return the gymnasium action space."""
        ...

    @abstractmethod
    def translate(self, action: Any, candle: dict, state: dict, strategy: Any) -> None:
        """Execute the action using strategy helpers (create_order, close_order, etc.)."""
        ...

    @abstractmethod
    def action_labels(self) -> List[str]:
        """Human-readable labels for each action (for logging/debugging)."""
        ...


# ---------------------------------------------------------------------------
# Discrete action spaces
# ---------------------------------------------------------------------------


@dataclass
class DiscreteActions(ActionSpace):
    """Configurable discrete action space.

    Default actions: ["hold", "open_long", "open_short", "close"]

    Supported action names:
      - "hold"       : do nothing
      - "open_long"  : open a long position at current close
      - "open_short" : open a short position at current close
      - "close"      : close current position at current close

    Parameters:
      actions: list of action name strings
      margin_pct: fraction of equity to use per trade (default 0.1 = 10%)
      sl_pct: stop-loss distance as fraction of entry price (0 = disabled)
      tp_pct: take-profit distance as fraction of entry price (0 = disabled)
    """

    actions: List[str] = field(
        default_factory=lambda: ["hold", "open_long", "open_short", "close"]
    )
    margin_pct: float = 0.1
    sl_pct: float = 0.0
    tp_pct: float = 0.0

    def gymnasium_space(self) -> gym.Space:
        return gym.spaces.Discrete(len(self.actions))

    def action_labels(self) -> List[str]:
        return list(self.actions)

    def translate(self, action: Any, candle: dict, state: dict, strategy: Any) -> None:
        if isinstance(action, np.ndarray):
            val = action.item() if action.size == 1 else action[0]
        elif isinstance(action, (list, tuple)):
            val = action[0]
        else:
            val = action

        idx = int(val)
        if idx < 0 or idx >= len(self.actions):
            return  # invalid -> hold

        name = self.actions[idx]
        close_px = float(candle.get("close", 0.0) or 0.0)
        pos = state.get("position")

        if name == "hold":
            return

        elif name == "open_long":
            if pos:
                return  # already in a position
            sl = close_px * (1.0 - self.sl_pct) if self.sl_pct > 0 else None
            tp = close_px * (1.0 + self.tp_pct) if self.tp_pct > 0 else None
            strategy.create_order(
                "LONG", price=close_px, margin_pct=self.margin_pct, sl=sl, tp=tp
            )

        elif name == "open_short":
            if pos:
                return
            sl = close_px * (1.0 + self.sl_pct) if self.sl_pct > 0 else None
            tp = close_px * (1.0 - self.tp_pct) if self.tp_pct > 0 else None
            strategy.create_order(
                "SHORT", price=close_px, margin_pct=self.margin_pct, sl=sl, tp=tp
            )

        elif name == "close":
            if not pos:
                return
            strategy.close_order(price=close_px)


@dataclass
class DiscreteActionsWithSizing(ActionSpace):
    """Discrete actions with multiple position sizes.

    Actions: hold, long_small, long_medium, long_large, short_small, short_medium, short_large, close

    Parameters:
      sizes: mapping of size label to margin_pct (e.g., {"small": 0.05, "medium": 0.1, "large": 0.2})
      sl_pct: stop-loss distance (0 = disabled)
      tp_pct: take-profit distance (0 = disabled)
    """

    sizes: Dict[str, float] = field(
        default_factory=lambda: {
            "small": 0.05,
            "medium": 0.1,
            "large": 0.2,
        }  # Small = 5% margin, Medium = 10%, Large = 20%
    )
    sl_pct: float = 0.0
    tp_pct: float = 0.0

    def __post_init__(self):
        self._actions: List[str] = ["hold"]
        self._margin_map: Dict[str, float] = {}
        for label, mpct in self.sizes.items():
            for side in ("long", "short"):
                name = f"{side}_{label}"
                self._actions.append(name)
                self._margin_map[name] = mpct
        self._actions.append("close")

    def gymnasium_space(self) -> gym.Space:
        return gym.spaces.Discrete(len(self._actions))

    def action_labels(self) -> List[str]:
        return list(self._actions)

    def translate(self, action: Any, candle: dict, state: dict, strategy: Any) -> None:
        if isinstance(action, np.ndarray):
            val = action.item() if action.size == 1 else action[0]
        elif isinstance(action, (list, tuple)):
            val = action[0]
        else:
            val = action

        idx = int(val)
        if idx < 0 or idx >= len(self._actions):
            return

        name = self._actions[idx]
        close_px = float(candle.get("close", 0.0) or 0.0)
        pos = state.get("position")

        if name == "hold":
            return
        elif name == "close":
            if pos:
                strategy.close_order(price=close_px)
            return

        # long_small, short_medium, etc.
        margin_pct = self._margin_map.get(name, 0.1)
        if pos:
            return  # already in position

        if name.startswith("long_"):
            sl = close_px * (1.0 - self.sl_pct) if self.sl_pct > 0 else None
            tp = close_px * (1.0 + self.tp_pct) if self.tp_pct > 0 else None
            strategy.create_order(
                "LONG", price=close_px, margin_pct=margin_pct, sl=sl, tp=tp
            )
        elif name.startswith("short_"):
            sl = close_px * (1.0 + self.sl_pct) if self.sl_pct > 0 else None
            tp = close_px * (1.0 - self.tp_pct) if self.tp_pct > 0 else None
            strategy.create_order(
                "SHORT", price=close_px, margin_pct=margin_pct, sl=sl, tp=tp
            )


# ---------------------------------------------------------------------------
# Continuous action space
# ---------------------------------------------------------------------------


@dataclass
class ContinuousActions(ActionSpace):
    """Continuous action space for PPO / SAC style agents.

    Action is a single float in [-1, 1]:
      - [-1, -threshold): open short with margin_pct = |action| * max_margin_pct
      - [-threshold, threshold]: hold (dead zone)
      - (threshold, 1]: open long with margin_pct = |action| * max_margin_pct

    If already in a position:
      - Action in opposite direction triggers close
      - Action in same direction is ignored (already positioned)

    Parameters:
      max_margin_pct: maximum margin percentage when |action| = 1.0
      threshold: dead zone around 0 where agent holds (default 0.1)
      sl_pct: stop-loss (0 = disabled)
      tp_pct: take-profit (0 = disabled)
    """

    max_margin_pct: float = 0.2
    threshold: float = 0.1
    sl_pct: float = 0.0
    tp_pct: float = 0.0

    def gymnasium_space(self) -> gym.Space:
        return gym.spaces.Box(low=-1.0, high=1.0, shape=(1,), dtype=np.float32)

    def action_labels(self) -> List[str]:
        return ["position_signal"]

    def translate(self, action: Any, candle: dict, state: dict, strategy: Any) -> None:
        if isinstance(action, np.ndarray):
            val = action.item() if action.size == 1 else action[0]
        elif isinstance(action, (list, tuple)):
            val = action[0]
        else:
            val = action

        signal = float(np.clip(val, -1.0, 1.0))

        close_px = float(candle.get("close", 0.0) or 0.0)
        pos = state.get("position")

        # Dead zone: do nothing (hold current state).
        # Positions exit via SL/TP or opposite-direction signal.
        if abs(signal) <= self.threshold:
            return

        if signal > self.threshold:
            # Want long
            if pos:
                if pos.get("side") == "SHORT":
                    strategy.close_order(price=close_px)
                return  # already long or just closed short
            margin = abs(signal) * self.max_margin_pct
            sl = close_px * (1.0 - self.sl_pct) if self.sl_pct > 0 else None
            tp = close_px * (1.0 + self.tp_pct) if self.tp_pct > 0 else None
            strategy.create_order(
                "LONG", price=close_px, margin_pct=margin, sl=sl, tp=tp
            )

        else:  # signal < -threshold
            # Want short
            if pos:
                if pos.get("side") == "LONG":
                    strategy.close_order(price=close_px)
                return  # already short or just closed long
            margin = abs(signal) * self.max_margin_pct
            sl = close_px * (1.0 + self.sl_pct) if self.sl_pct > 0 else None
            tp = close_px * (1.0 - self.tp_pct) if self.tp_pct > 0 else None
            strategy.create_order(
                "SHORT", price=close_px, margin_pct=margin, sl=sl, tp=tp
            )

File: src/rl/actions.rs
//! RL Action Spaces — Phase 6
//!
//! Translates f64 action values from Gymnasium into native engine order calls.
//! Ports: DiscreteActions, DiscreteActionsWithSizing, ContinuousActions.

use crate::engine::RustTradingEngine;
use crate::types::*;

// ---------------------------------------------------------------------------
// RlActionSpace — enum dispatching to the three concrete spaces
// ---------------------------------------------------------------------------

pub enum RlActionSpace {
    Discrete(DiscreteActionSpace),
    DiscreteWithSizing(DiscreteWithSizingSpace),
    Continuous(ContinuousActionSpace),
}

impl RlActionSpace {
    /// Translate a raw f64 action into engine order calls.
    /// `close_price` is extracted from the current candle close.
    pub fn translate(&self, engine: &mut RustTradingEngine, action_val: f64, close_price: f64) {
        match self {
            Self::Discrete(d) => d.translate(engine, action_val, close_price),
            Self::DiscreteWithSizing(d) => d.translate(engine, action_val, close_price),
            Self::Continuous(c) => c.translate(engine, action_val, close_price),
        }
    }

    /// Number of discrete actions (for Gymnasium space definition).
    pub fn num_actions(&self) -> usize {
        match self {
            Self::Discrete(d) => d.actions.len(),
            Self::DiscreteWithSizing(d) => d.actions.len(),
            Self::Continuous(_) => 1, // Box(1,)
        }
    }
}

// ---------------------------------------------------------------------------
// DiscreteActionSpace
// ---------------------------------------------------------------------------

/// Action names for discrete space.
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum DiscreteAction {
    Hold,
    OpenLong,
    OpenShort,
    Close,
}

impl DiscreteAction {
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "hold" => Some(Self::Hold),
            "open_long" => Some(Self::OpenLong),
            "open_short" => Some(Self::OpenShort),
            "close" => Some(Self::Close),
            _ => None,
        }
    }
}

pub struct DiscreteActionSpace {
    pub actions: Vec<DiscreteAction>,
    pub margin_pct: f64,
    pub sl_pct: f64,
    pub tp_pct: f64,
}

impl DiscreteActionSpace {
    pub fn new(action_names: &[&str], margin_pct: f64, sl_pct: f64, tp_pct: f64) -> Self {
        let actions: Vec<DiscreteAction> = action_names
            .iter()
            .filter_map(|s| DiscreteAction::from_str(s))
            .collect();
        Self {
            actions,
            margin_pct,
            sl_pct,
            tp_pct,
        }
    }

    pub fn default() -> Self {
        Self::new(
            &["hold", "open_long", "open_short", "close"],
            0.1,
            0.0,
            0.0,
        )
    }

    fn translate(&self, engine: &mut RustTradingEngine, action_val: f64, close_price: f64) {
        let idx = action_val as usize;
        if idx >= self.actions.len() {
            return; // invalid -> hold
        }
        let action = self.actions[idx];
        let has_position = engine.position.is_some();

        match action {
            DiscreteAction::Hold => {}
            DiscreteAction::OpenLong => {
                if has_position {
                    return;
                }
                let sl = if self.sl_pct > 0.0 {
                    Some(close_price * (1.0 - self.sl_pct))
                } else {
                    None
                };
                let tp = if self.tp_pct > 0.0 {
                    Some(close_price * (1.0 + self.tp_pct))
                } else {
                    None
                };
                engine.create_order_internal(
                    OrderSide::LONG,
                    close_price,
                    self.margin_pct,
                    tp,
                    sl,
                );
            }
            DiscreteAction::OpenShort => {
                if has_position {
                    return;
                }
                let sl = if self.sl_pct > 0.0 {
                    Some(close_price * (1.0 + self.sl_pct))
                } else {
                    None
                };
                let tp = if self.tp_pct > 0.0 {
                    Some(close_price * (1.0 - self.tp_pct))
                } else {
                    None
                };
                engine.create_order_internal(
                    OrderSide::SHORT,
                    close_price,
                    self.margin_pct,
                    tp,
                    sl,
                );
            }
            DiscreteAction::Close => {
                if !has_position {
                    return;
                }
                engine.close_order_internal(close_price);
            }
        }
    }
}

// ---------------------------------------------------------------------------
// DiscreteWithSizingSpace
// ---------------------------------------------------------------------------

/// An action entry for discrete-with-sizing: side + margin_pct, or Hold/Close.
#[derive(Clone, Debug)]
pub enum SizedAction {
    Hold,
    OpenLong(f64),  // margin_pct
    OpenShort(f64), // margin_pct
    Close,
}

pub struct DiscreteWithSizingSpace {
    pub actions: Vec<SizedAction>,
    pub sl_pct: f64,
    pub tp_pct: f64,
}

impl DiscreteWithSizingSpace {
    /// Build from a list of (label, margin_pct) pairs, e.g. [("small", 0.05), ("medium", 0.1)].
    pub fn new(sizes: &[(&str, f64)], sl_pct: f64, tp_pct: f64) -> Self {
        let mut actions = vec![SizedAction::Hold];
        for (_label, mpct) in sizes {
            actions.push(SizedAction::OpenLong(*mpct));
            actions.push(SizedAction::OpenShort(*mpct));
        }
        actions.push(SizedAction::Close);
        Self {
            actions,
            sl_pct,
            tp_pct,
        }
    }

    pub fn default() -> Self {
        Self::new(
            &[("small", 0.05), ("medium", 0.1), ("large", 0.2)],
            0.0,
            0.0,
        )
    }

    fn translate(&self, engine: &mut RustTradingEngine, action_val: f64, close_price: f64) {
        let idx = action_val as usize;
        if idx >= self.actions.len() {
            return;
        }
        let has_position = engine.position.is_some();

        match &self.actions[idx] {
            SizedAction::Hold => {}
            SizedAction::Close => {
                if has_position {
                    engine.close_order_internal(close_price);
                }
            }
            SizedAction::OpenLong(margin_pct) => {
                if has_position {
                    return;
                }
                let sl = if self.sl_pct > 0.0 {
                    Some(close_price * (1.0 - self.sl_pct))
                } else {
                    None
                };
                let tp = if self.tp_pct > 0.0 {
                    Some(close_price * (1.0 + self.tp_pct))
                } else {
                    None
                };
                engine.create_order_internal(OrderSide::LONG, close_price, *margin_pct, tp, sl);
            }
            SizedAction::OpenShort(margin_pct) => {
                if has_position {
                    return;
                }
                let sl = if self.sl_pct > 0.0 {
                    Some(close_price * (1.0 + self.sl_pct))
                } else {
                    None
                };
                let tp = if self.tp_pct > 0.0 {
                    Some(close_price * (1.0 - self.tp_pct))
                } else {
                    None
                };
                engine.create_order_internal(OrderSide::SHORT, close_price, *margin_pct, tp, sl);
            }
        }
    }
}

// ---------------------------------------------------------------------------
// ContinuousActionSpace
// ---------------------------------------------------------------------------

pub struct ContinuousActionSpace {
    pub max_margin_pct: f64,
    pub threshold: f64,
    pub sl_pct: f64,
    pub tp_pct: f64,
}

impl ContinuousActionSpace {
    pub fn new(max_margin_pct: f64, threshold: f64, sl_pct: f64, tp_pct: f64) -> Self {
        Self {
            max_margin_pct,
            threshold,
            sl_pct,
            tp_pct,
        }
    }

    pub fn default() -> Self {
        Self::new(0.2, 0.1, 0.0, 0.0)
    }

    fn translate(&self, engine: &mut RustTradingEngine, action_val: f64, close_price: f64) {
        let signal = action_val.clamp(-1.0, 1.0);
        let has_position = engine.position.is_some();

        // Dead zone: do nothing (hold current state).
        // Positions exit via SL/TP or opposite-direction signal.
        if signal.abs() <= self.threshold {
            return;
        }

        if signal > self.threshold {
            // Want long
            if has_position {
                let is_short = engine
                    .position
                    .as_ref()
                    .map_or(false, |p| p.side == PositionSide::SHORT);
                if is_short {
                    engine.close_order_internal(close_price);
                }
                return; // already long or just closed short
            }
            let margin = signal.abs() * self.max_margin_pct;
            let sl = if self.sl_pct > 0.0 {
                Some(close_price * (1.0 - self.sl_pct))
            } else {
                None
            };
            let tp = if self.tp_pct > 0.0 {
                Some(close_price * (1.0 + self.tp_pct))
            } else {
                None
            };
            engine.create_order_internal(OrderSide::LONG, close_price, margin, tp, sl);
        } else {
            // signal < -threshold -> want short
            if has_position {
                let is_long = engine
                    .position
                    .as_ref()
                    .map_or(false, |p| p.side == PositionSide::LONG);
                if is_long {
                    engine.close_order_internal(close_price);
                }
                return; // already short or just closed long
            }
            let margin = signal.abs() * self.max_margin_pct;
            let sl = if self.sl_pct > 0.0 {
                Some(close_price * (1.0 + self.sl_pct))
            } else {
                None
            };
            let tp = if self.tp_pct > 0.0 {
                Some(close_price * (1.0 - self.tp_pct))
            } else {
                None
            };
            engine.create_order_internal(OrderSide::SHORT, close_price, margin, tp, sl);
        }
    }
}
